pkg/controller/utils/utils.go

```diff
@@ -986,3 +986,53 @@ func RemoveInstallationFinalizer(i *operatorv1.Installation, finalizer string) {
 		i.SetFinalizers(stringsutil.RemoveStringInSlice(finalizer, i.GetFinalizers()))
 	}
 }
+
+// MaintainInstallationFinalizer manages a controller's finalizer on the Installation resource.
+// We add a finalizer to the Installation when the mainResource has been installed, and only remove that finalizer when
+// the resource has been deleted and its secondary resources have stopped running. This allows for a graceful cleanup of any resources
+// prior to the CNI plugin being removed.
+func MaintainInstallationFinalizer(
+	ctx context.Context,
+	c client.Client,
+	mainResource client.Object,
+	finalizer string,
+	secondaryResources ...client.Object,
+) error {
+	// Get the Installation.
+	installation := &operatorv1.Installation{}
+	if err := c.Get(ctx, DefaultInstanceKey, installation); err != nil {
+		if errors.IsNotFound(err) {
+			log.V(1).Info("Installation config not found")
+			return nil
+		}
+		log.Error(err, "An error occurred when querying the Installation resource")
+		return err
+	}
+	patchFrom := client.MergeFrom(installation.DeepCopy())
+
+	// Determine the correct finalizers to apply to the Installation.
+	if mainResource != nil {
+		// Add a finalizer indicating that the mainResource is still available.
+		SetInstallationFinalizer(installation, finalizer)
+	} else {
+		// Check if the namespaced secondaryResources are still present.
+		// Keep track of all the secondary resources that the main resource creates.
+		// Only delete the finalizer if all of the secondary resources are deleted.
+		for _, secondaryResource := range secondaryResources {
+			err := c.Get(ctx, types.NamespacedName{Namespace: secondaryResource.GetNamespace(), Name: secondaryResource.GetName()}, secondaryResource)
+			if err != nil && !errors.IsNotFound(err) {
+				return err
+			} else if errors.IsNotFound(err) {
+				log.Info("Object no longer exists.", secondaryResource.GetName(), secondaryResource)
```

Comment:
```suggestion
				log.Info("Object no longer exists.", "object", secondaryResource)
```

Intention of these loggers is to produce key=value pairs that can be searched - the name should already be present from `secondaryResource` so I don't think we need it in both the key and the value?